### PR TITLE
Added support for INameOfOperation to the control flow graph

### PR DESCRIPTION
**Pull request by chborl**
Tagging @333fred  for review.

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

**Review by mavasani — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -3572,6 +3572,11 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
             return Operation.CreateOperationNone(semanticModel: null, operation.Syntax, operation.ConstantValue, childrenBuilder.ToImmutableAndFree(), operation.IsImplicit);
         }
 
+        public override IOperation VisitNameOf(INameOfOperation operation, int? captureIdForResult)
+        {
+            return new NameOfExpression(Visit(operation.Argument), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >Visit(operation.Argument) [](start = 40, length = 25)

I think we should simply replace this node with a string literal that it evaluates to. Visiting the argument is not the right thing to do, no matter what is there, it is not going to be executed and should any control flow be inside this expression, this should not have any effect on the flow graph.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -3572,6 +3572,12 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
             return Operation.CreateOperationNone(semanticModel: null, operation.Syntax, operation.ConstantValue, childrenBuilder.ToImmutableAndFree(), operation.IsImplicit);
         }
 
+        public override IOperation VisitNameOf(INameOfOperation operation, int? captureIdForResult)
+        {
+            Debug.Assert(operation.ConstantValue.HasValue);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >operation.ConstantValue.HasValue [](start = 25, length = 32)

Is this assert going to hold for error cases?

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: @chborl looks like there's some case that this won't hold for VB: https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb#L692-L694. The field isn't allowed to be null in C#.

Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Will this not have a ConstantValue with `null` as the value, which means the assert still holds?

Comment (chborl) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Yes, I verified this with VB test NameOfFlow_01 which is an error case.  ConstantValue = null and HasValue is still true.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INameOfOperation.cs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        [CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)]
+        [Fact]
+        public void NameOfFlow_01()
+        {
+            string source = @"
+class C
+{
+    void M(bool b, int i1, int i2)
+    /*<bind>*/{
+        string test = nameof(b ? i1 : i2);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INameOfOperation.cs]: >string test [](start = 8, length = 11)

In the future CFG PRs, consider avoiding declaring locals unless you are specifically testing an aspect related to declaring locals. The reason is that creates regions in the graph and somewhat bloats base-lines, creates more stuff to process for a human.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INameOfOperation.cs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    }/*</bind>*/
+}
+";
+            var expectedDiagnostics = new DiagnosticDescription[] {
+                // CS8081: Expression does not have a name.
+                //         string test = nameof(b ? i1 : i2);
+                Diagnostic(ErrorCode.ERR_ExpressionHasNoName, "b ? i1 : i2").WithLocation(6, 30),
+                // CS0219: The variable 'test' is assigned but its value is never used
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INameOfOperation.cs]: >// CS0219: The variable 'test' is assigned but its value is never used [](start = 16, length = 70)

Also creates diagnostics like that, which is not a problem when there are other errors, but could be annoying for success cases

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 2)

**Pull request merged**